### PR TITLE
Authentication

### LINKED ISSUE #11: Authentication
#### As an application hoster I want to add user authentication to my application.

- [x] Users can sign up.
- [x] Users can log in after they've signed up.
- [x] Users can log out after logging in.
- [x] Users can choose to be remembered after logging in.
- [x] Multiple browsers can concurrently hold permanent cookies.
- [x] Logged in users should not be able to register, nor log in again.

__Note__: Do not use Devise for this task.

Discussion topics:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

Reference material:
- https://www.railstutorial.org/book/log_in_log_out

Further reading:
- https://github.com/plataformatec/devise

Dependencies:
- Delete Image

### PR DESCRIPTION
**Pull request by Xin-Tong**
Closes #11

### CODE REVIEW

**Comment by bboe** [comment]
> Multiple browsers can concurrently hold permanent cookies.

This requirement is not satisfied. Steps to reproduce:

1) Open firefox, login as a user with "remember password" selected
2) Open chrome, login as the same user with "remember password" selected
3) Close firefox
4) Reopen firefox and observe that the user is not still logged in

Note that if you skipped step 2 the user would still be logged in.

**Comment by Xin-Tong** [comment]
@bboe I squashed the modification to remove token from database into https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/284520e7324b7f3bf90895f333a945f795d489ce since it's better to just delete the unused migration.

**Comment by bboe** [comment]
Great. We'll see what @d-ray has to say.

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+include SessionsHelper
+
+class SessionsController < ApplicationController
+  def new
+    redirect_to root_path,
+                flash: {
+                  warning: "Hi #{@current_user.name}, you've already logged in."
+                } if logged_in?
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Do you think that an inline `if` after a multi-line method call promotes optimal readability?

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: What will happen if a user is already logged in and tries to re-authenticate?

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      render :new
+    end
+  end
+
+  def destroy
+    if logged_in?
+      log_out!
+      redirect_to login_path
+    elsif User.find_by(id: params[:id])
+      redirect_to root_path, flash: { danger: 'You have already logged out.' }
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Does having different responses based on whether the id is valid expose any sensitive information?

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (Xin-Tong) [inline app/controllers/sessions_controller.rb]: Good suggestion. I'll address it later.

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (Xin-Tong) [inline app/controllers/sessions_controller.rb]: Do you mean the flash of Line7? https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32#discussion-diff-61662695R7

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (Xin-Tong) [inline app/controllers/sessions_controller.rb]: User is searched by user ID instead of user email. So ID here is just a number, shouldn't have security issue.

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Right, line 7 handles this case in the `new` action.  My question is more about what happens for the same situation in the `create` action.

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      flash[:success] = 'Welcome!'
+      redirect_to login_path
+    else
+      render :new, status: :unprocessable_entity
```
Comment (d-ray) [inline app/controllers/users_controller.rb]: Should there be any indication to the user of what specifically went wrong?

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    end
+  end
+
+  def index
+    @users = User.all
```
Comment (d-ray) [inline app/controllers/users_controller.rb]: What's the purpose of having an `index` action for users?

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:name, :email, :password)
```
Comment (d-ray) [inline app/controllers/users_controller.rb]: What will happen if `user` isn't included in the params?

File: app/helpers/navbar_helper.rb [hunk 7]
```diff
@@ -1,9 +1,9 @@
 module NavbarHelper
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text:, path:, is_active:, html_options: 'get')
```
Comment (d-ray) [inline app/helpers/navbar_helper.rb]: Is `html_options` an accurate name given how this parameter is actually used?

File: app/views/layouts/_navbar.html.erb [hunk 8]
```diff
@@ -3,6 +3,14 @@
     <ul class="nav navbar-nav">
       <%= nav_item(text: 'Index', path: root_path, is_active: current_page?(images_path) || current_page?(root_path)) %>
       <%= nav_item(text: 'New', path: new_image_path, is_active: current_page?(new_image_path)) %>
+      <div class="pull-xs-right">
+        <%= nav_item(text: 'Sign up now!', path: signup_path, is_active: current_page?(signup_path)) %>
```
Comment (d-ray) [inline app/views/layouts/_navbar.html.erb]: It seems like the logic for which nav item should be active kind of clutters up this partial a bit.  Are there any cleaner ways to handle this?

File: app/views/users/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<% @users.each do |user| %>
+  <ul>
+    <li>
+      <%= "Name: #{user.name} Email: #{user.email}" %>
+      </li>
```
Comment (d-ray) [inline app/views/users/index.html.erb]: The indentation is off here.

File: app/views/users/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+<% @users.each do |user| %>
+  <ul>
```
Comment (d-ray) [inline app/views/users/index.html.erb]: Do you want the `ul` element inside of the loop?

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
+include SessionsHelper
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Do you need all of the methods provided by the helper?

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    elsif User.find_by(id: params[:id])
+      redirect_to root_path, flash: { danger: 'You have already logged out.' }
+    else
+      redirect_to root_path, flash: { danger: 'Please sign up first.' }
+    end
+  end
+
+  private
+
+  def log_in!(user)
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: This is also defined in the included `SessionsHelper`.

File: app/helpers/sessions_helper.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+        @current_user = user
+      end
+    end
+  end
+
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def log_in!(user)
```
Comment (d-ray) [inline app/helpers/sessions_helper.rb]: Does it make sense to expose a method like this in the view by including it in the helper?

File: app/helpers/sessions_helper.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
+module SessionsHelper
```
Comment (d-ray) [inline app/helpers/sessions_helper.rb]: Is there a corresponding test for this helper?

File: test/controllers/sessions_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  test 'log in with invalid information' do
+    post :create, session: { email: 'test@email.com', password: 'wrong password' }
+
+    refute is_logged_in?(@user)
+    assert_select "form[action='#{login_path}']", 1
+    assert_equal 'Invalid email/password combination', flash[:danger]
+  end
+
+  test 'user is logged_in using remember me' do
+    cookies.permanent.signed[:user_id] = @user.id
```
Comment (d-ray) [inline test/controllers/sessions_controller_test.rb]: Does this test provide value in its current form?

File: test/models/user_test.rb [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    @user.email = mixed_case_email
+    @user.save!
+
+    assert_equal mixed_case_email.downcase, @user.reload.email
+  end
+
+  test 'email addresses should be unique' do
+    duplicate_user = User.new(name: 'Duplicate User', email: 'user@example.com', password: 'something else')
+    duplicate_user.email = @user.email.upcase
+    @user.save
```
Comment (d-ray) [inline test/models/user_test.rb]: Do you need to save the user here?

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (Xin-Tong) [inline app/controllers/users_controller.rb]: 👍   I will add a flash message.

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (Xin-Tong) [inline app/controllers/users_controller.rb]: Almost just for double-check at that time. I forgot to delete. Thanks.

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (Xin-Tong) [inline app/controllers/users_controller.rb]: It will raise ```ActionController::ParameterMissing: param is missing or the value is empty: user``` since we use `require` here. But I don't know how to reproduce it in browser. Is it an issue we need to handle?

File: app/helpers/navbar_helper.rb (hunk 7 above)
Comment (Xin-Tong) [inline app/helpers/navbar_helper.rb]: Maybe rename it as `http_verb` ?

File: app/views/layouts/_navbar.html.erb (hunk 8 above)
Comment (Xin-Tong) [inline app/views/layouts/_navbar.html.erb]: Will address it.

File: app/helpers/sessions_helper.rb (hunk 13 above)
Comment (Xin-Tong) [inline app/helpers/sessions_helper.rb]: I've thought about where to put this method. But I need to use it in the `current_user` . On the other hand, I need to use the `log_in` in controller test. So I chose to put it here and in the test.

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (d-ray) [inline app/controllers/users_controller.rb]: If you're satisfied with the way Rails handles it, then there's probably no need to customize.

File: app/helpers/navbar_helper.rb (hunk 7 above)
Comment (d-ray) [inline app/helpers/navbar_helper.rb]: That would be a more accurate name.  You could also make `html_options` a hash and read the verb out of it, e.g.

    def nav_item(text:, path:, is_active:, html_options: {})
      http_verb = html_options[:http_verb] || 'get'
      …
    end

but if `http_verb` is the only option you want to support it probably makes more sense to do what you suggested.

File: app/helpers/sessions_helper.rb (hunk 13 above)
Comment (d-ray) [inline app/helpers/sessions_helper.rb]: > I need to use it in the `current_user`

It looks like `current_user` is only called by `logged_in?` (since the controller references `@current_user` directly); `logged_in?` in turn is only called from the view.  As a more general follow up to my original question, does it make sense for the action of logging in a user to happen (or not) based on method calls in the view?  If not, what is the "right" place to take that action?

Comment (d-ray) [inline app/helpers/sessions_helper.rb]: On a totally separate note, what was your reasoning for naming the method with a `!` (i.e. `log_in!` as opposed to `log_in`)?

File: app/helpers/navbar_helper.rb (hunk 7 above)
Comment (Xin-Tong) [inline app/helpers/navbar_helper.rb]: Sounds good. I will do it. Thanks for elaborating.

File: app/helpers/sessions_helper.rb (hunk 13 above)
Comment (Xin-Tong) [inline app/helpers/sessions_helper.rb]: `logged_in?` and `current_user` are used in every view since they're called from nav bar. Does it make sense to move they two into a module called `helpers/common_helpers/login_helper.rb` then include this helper into `application_helper`?

Comment (Xin-Tong) [inline app/helpers/sessions_helper.rb]: Can we discuss about it tomorrow? I'm thinking about use `before_action` but not very sure how to separate.

File: app/views/users/index.html.erb (hunk 10 above)
Comment (Xin-Tong) [inline app/views/users/index.html.erb]: I will pull it out of the loop.

File: app/views/users/index.html.erb (hunk 9 above)
Comment (Xin-Tong) [inline app/views/users/index.html.erb]: will address.

File: test/controllers/sessions_controller_test.rb (hunk 15 above)
Comment (Xin-Tong) [inline test/controllers/sessions_controller_test.rb]: I should `get :new` here. will address it.

File: test/models/user_test.rb (hunk 16 above)
Comment (Xin-Tong) [inline test/models/user_test.rb]: You're right. I originally just want to check if it's valid. I will remove it.

Comment (d-ray) [inline test/models/user_test.rb]: I think I'm wrong, actually, but try it and see what happens.

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (Xin-Tong) [inline app/controllers/users_controller.rb]: I am addressing this comment now and realize there are validations according to different input fields (so the according error messages will show as well). Is it necessary to add an extra error message here?

File: app/views/layouts/_navbar.html.erb (hunk 8 above)
Comment (Xin-Tong) [inline app/views/layouts/_navbar.html.erb]: In Line4 we have `is_active: current_page?(images_path) || current_page?(root_path)`, I am not sure how to refactor this part since there are two inputs.

Comment (d-ray) [inline app/views/layouts/_navbar.html.erb]: How would you refactor it if there was only one input on each line?

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (d-ray) [inline app/controllers/users_controller.rb]: Ah, right.  If there are inline form errors you probably don't need a flash message as well.

File: app/views/layouts/_navbar.html.erb (hunk 8 above)
Comment (Xin-Tong) [inline app/views/layouts/_navbar.html.erb]: move the `current_page?(signup_path)` into `nav_item`, so inside `nav_item`, we have something like: ``` is_active = current_page?(path)```

Comment (d-ray) [inline app/views/layouts/_navbar.html.erb]: Ok, great.  That sounds good; I would probably recommend naming the parameter something that better reflects it's usage though.  It _is_ a path, but in this context it _represents_ the page on which the link should be active.  You might call it `active_page`.  Now, the only difference between what you suggested here and what you need (to handle line 4) is that there can be multiple active pages.  How can you adjust what you have to meet that requirement?

File: test/models/user_test.rb (hunk 16 above)
Comment (Xin-Tong) [inline test/models/user_test.rb]: I double check that we need the `save` here. Thanks for making me to think twice about it.

File: app/views/layouts/_navbar.html.erb (hunk 8 above)
Comment (Xin-Tong) [inline app/views/layouts/_navbar.html.erb]: Can I do it like this?
```
module NavbarHelper
  def nav_item(text:, active_page:, http_verb: 'get')
    is_active = active_page.any? {|path| current_page?(path)}
    path = active_page[0]
    classes = %w(nav-item)
    classes << 'active' if is_active
    content_tag :li, class: classes do
      link_to text, path, class: 'nav-link', method: http_verb
    end
  end
end
```
`active_page` will be an array.
But I don't like the line of ` path = active_page[0]` that much.

**Comment by Xin-Tong** [comment]
@d-ray I've addressed all the comments except this one https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32#discussion-diff-61782351R7. I haven't found a optimal way yet. Could you give some input?

File: app/views/layouts/_navbar.html.erb (hunk 8 above)
Comment (d-ray) [inline app/views/layouts/_navbar.html.erb]: > I would probably recommend naming the parameter something that better reflects it's usage though. It is a path, but in this context it represents the page on which the link should be active.

So I somehow overlooked the fact that the path parameter is also being used as the path to the link - `path` is a totally reasonable parameter name.  I agree that `path = active_page[0]` isn't great.  It seems like the issue with it is that it doesn't clearly distinguish between the link path and paths for other pages on which the link should be active.  It would be better to distinguish between these different concerns with a method signature like `def nav_item(text:, page:, other_active_pages: [], http_verb: 'get')`.  However, I don't think you actually need to support this.  What is `root_path` for the application?

Comment (Xin-Tong) [inline app/views/layouts/_navbar.html.erb]: The `root_path` is `/` which will hit the `images#index` action. We set it in `routes.rb`.

Comment (d-ray) [inline app/views/layouts/_navbar.html.erb]: Oh, of course.  It's a different path even though it's routed to the same place.  Ok, well you could try what I suggested above.  I think it's an improvement, but perhaps the improvement it offers doesn't justify the effort of making the change.  It's up to you how you want to handle this.

Comment (Xin-Tong) [inline app/views/layouts/_navbar.html.erb]: As long as there is improvement :) This is the commit https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/713dfa8a7ce46090e27d67ed9c54c3ea3f679def

**Comment by Xin-Tong** [comment]
@d-ray All the comments have been addressed. Please let me know if it's good to squash then merge.

File: app/controllers/sessions_controller.rb [hunk 17]
```diff
@@ -0,0 +1,57 @@
+include ApplicationHelper
+include LoginService
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: Should we be including `LoginService` here?

File: app/helpers/navbar_helper.rb [hunk 18]
```diff
@@ -1,9 +1,11 @@
 module NavbarHelper
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text:, path:, http_verb: 'get', additional_active_path: [])
+    all_active_path = additional_active_path << path
```
Comment (d-ray) [inline app/helpers/navbar_helper.rb]: Using `<<` here will modify the value of `additional_active_path`, which wouldn't necessarily be obvious to someone modifying this code in the future.  It might also be good to pluralize `additional_active_path` since it's an array and thus in theory could be many paths.

File: app/models/user.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
+class User < ActiveRecord::Base
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: 50 }
+  validates :email, presence: true, length: { maximum: 255 }, email_format: true, uniqueness: { case_sensitive: false }
```
Comment (d-ray) [inline app/models/user.rb]: You don't necessarily have to do this, but you could remove the `case_sensitive` option if you changed `before_save { self.email = email.downcase }` to use a `before_validation` hook instead.  See http://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html for more info on the different callbacks and the order in which they run.

File: test/controllers/sessions_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+    refute is_logged_in?(@user)
+    delete :destroy, id: @user.id
+
+    assert_redirected_to root_path
+    assert_equal 'You have already logged out.', flash[:danger]
+  end
+
+  private
+
+  def log_in!
```
Comment (d-ray) [inline test/controllers/sessions_controller_test.rb]: Should this method end in a `!`?  See http://dablog.rubypal.com/2007/8/15/bang-methods-or-danger-will-rubyist for an explanation of ruby conventions around methods ending in `!`.

File: app/helpers/navbar_helper.rb (hunk 18 above)
Comment (rmacklin) [inline app/helpers/navbar_helper.rb]: 👍 Exactly. To expand on that point, generally, you should avoid modifying objects passed as parameters whenever possible. Doing that is prone to causing unintended side effects (that can turn up as corner case bugs), but it can usually be easily avoided. In this case you can use array concatenation instead:
```ruby
all_active_paths = additional_active_paths + [path]
```

File: test/flow_tests/users_flow_test.rb [hunk 21]
```diff
@@ -0,0 +1,20 @@
+require 'flow_test_helper'
+
+class UsersFlowTest < FlowTestCase
+  test 'user flow' do
+    images_index_page = PageObjects::Images::IndexPage.visit
+
+    new_user_page = images_index_page.add_new_user!
```
Comment (d-ray) [inline test/flow_tests/users_flow_test.rb]: In light of https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/files#r63780576, I should probably also mention that we have a page object convention that contradicts the ruby convention:  https://github.com/appfolio/ae_page_objects#conventions.  Thus it's ok that `add_new_user!` here (and other page objects methods) end in `!` even though this is against the ruby convention.  Our convention supersedes the ruby convention in the context of page objects/flow tests, so within this context we know that `!` has a particular meaning.  Outside of the context of page objects/flow tests we should follow the ruby convention whenever possible.  In both cases, following the convention allows us to better communicate our intent to future readers of our code.

File: test/helpers/navbar_helper_test.rb [hunk 22]
```diff
@@ -1,14 +1,20 @@
 require 'test_helper'
 
 class NavbarHelperTest < ActionView::TestCase
+  attr_reader :request
```
Comment (d-ray) [inline test/helpers/navbar_helper_test.rb]: I think I may have mentioned some benefits of using `attr_reader` over directly referencing instance variables when we paired.  In my mind, those benefits don't really apply to tests - I think this is more likely to confuse someone by making them think that calling code somewhere will be accessing `request` on this class.  What benefits do you see from using it here?

File: test/models/user_test.rb (hunk 16 above)
Comment (d-ray) [inline test/models/user_test.rb]: Sure. 😃  In my original comment I assumed that the user was saved on creation, but after writing it I realized that it wasn't.  I chose to leave the comment just so that you would have to think twice.

**Comment by d-ray** [comment]
@Xin-Tong I took another look and left a few more comments.  Let me know when you've reviewed and/or addressed them.

File: app/controllers/sessions_controller.rb (hunk 17 above)
Comment (Xin-Tong) [inline app/controllers/sessions_controller.rb]: It seems not. I will remove it.

File: app/helpers/navbar_helper.rb (hunk 18 above)
Comment (Xin-Tong) [inline app/helpers/navbar_helper.rb]: Great suggestion. I'll change it.

File: app/models/user.rb (hunk 19 above)
Comment (Xin-Tong) [inline app/models/user.rb]: Nice. Thank you for the information link.

File: test/controllers/sessions_controller_test.rb (hunk 20 above)
Comment (Xin-Tong) [inline test/controllers/sessions_controller_test.rb]: I changed the one in `LoginServe` but forgot this one. I'll address it.

File: test/flow_tests/users_flow_test.rb (hunk 21 above)
Comment (Xin-Tong) [inline test/flow_tests/users_flow_test.rb]: Yes, I try to follow our convention in all the `!` case in this project ("change current window"). Functionally, after log in, we redirect the user to the index page. So I considered it as a "change window". But as you said before, it against the ruby convention, and the log_in function alone do not necessarily redirect the users. So it makes sense to remove the `!` there. I would prefer to keep `!` here.

File: test/helpers/navbar_helper_test.rb (hunk 22 above)
Comment (Xin-Tong) [inline test/helpers/navbar_helper_test.rb]: I cannot get access to request variable without this getter but I want to use `request.path = root_path` in the following test to define the current page.

Comment (Xin-Tong) [inline test/helpers/navbar_helper_test.rb]: Or should I stub current_page?

File: test/flow_tests/users_flow_test.rb (hunk 21 above)
Comment (d-ray) [inline test/flow_tests/users_flow_test.rb]: Sorry, I wasn't suggesting that you remove the `!` here.  Rather, I was pointing out that the ruby convention I mentioned in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/files#r63780576 doesn't apply in this case.  I was trying to avoid confusion, but I may have caused more confusion than if I just hadn't said anything.  Oops 😞

File: test/helpers/navbar_helper_test.rb (hunk 22 above)
Comment (d-ray) [inline test/helpers/navbar_helper_test.rb]: > I cannot get access to request variable without this getter
`attr_reader` should just be wrapping an instance variable of the same name.  Does `@request` not work here?

File: test/flow_tests/users_flow_test.rb (hunk 21 above)
Comment (Xin-Tong) [inline test/flow_tests/users_flow_test.rb]: I misunderstood your message. Now it makes more sense for both `!` cases. Thanks! 👍

File: test/helpers/navbar_helper_test.rb (hunk 22 above)
Comment (Xin-Tong) [inline test/helpers/navbar_helper_test.rb]: Then it will raise `Minitest::UnexpectedError: RuntimeError: You cannot use helpers that need to determine the current page unless your view context provides a Request object in a #request method`

Comment (d-ray) [inline test/helpers/navbar_helper_test.rb]: Interesting.  I think stubbing `current_page?` is probably a better solution over trying to set the path on the request.

Comment (Xin-Tong) [inline test/helpers/navbar_helper_test.rb]: Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/5f4c1ee0b0328dda2cf06f22b6900e849a9d266d and https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/3cebd634bc98dd21af4ef9b5ce7997bcc6d34aae

File: test/controllers/sessions_controller_test.rb (hunk 20 above)
Comment (Xin-Tong) [inline test/controllers/sessions_controller_test.rb]: Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

File: app/models/user.rb (hunk 19 above)
Comment (Xin-Tong) [inline app/models/user.rb]: Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

File: app/controllers/sessions_controller.rb (hunk 17 above)
Comment (Xin-Tong) [inline app/controllers/sessions_controller.rb]: Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

**Comment by Xin-Tong** [comment]
@d-ray I've addressed all the comments.

File: test/controllers/sessions_controller_test.rb (hunk 20 above)
Comment (Xin-Tong) [inline test/controllers/sessions_controller_test.rb]: Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

File: app/controllers/sessions_controller.rb [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+  def logged_in_redirect
+    if logged_in?
+      redirect_to root_path,
+                  flash: {
+                    warning: "Hi #{@current_user.name}, you've already logged in."
+                  }
+    end
+  end
+
+  def log_out!
```
Comment (d-ray) [inline app/controllers/sessions_controller.rb]: This method name also violates the ruby convention for `!` methods.

File: app/helpers/navbar_helper.rb [hunk 24]
```diff
@@ -1,9 +1,11 @@
 module NavbarHelper
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text:, path:, http_verb: 'get', additional_active_paths: [])
+    all_active_paths = additional_active_paths + [path]
+    is_active = all_active_paths.any? {|path| current_page?(path)}
```
Comment (d-ray) [inline app/helpers/navbar_helper.rb]: Not super important, but you could inline this where it's used on line 6 instead of setting it to a variable.

File: config/routes.rb [hunk 25]
```diff
@@ -5,6 +5,13 @@
     get 'new_share', on: :member
     post 'create_share', on: :member
   end
+  
+  get 'signup' => 'users#new'
+  resources :users, only: [:index, :new, :create]
```
Comment (d-ray) [inline config/routes.rb]: You previously removed the index action from the users controller, but the route is still here.

File: test/models/user_test.rb [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+      @user.email = valid_address
+      assert @user.valid?, "#{valid_address.inspect} should be valid"
+    end
+  end
+
+  test 'email validation should reject invalid addresses' do
+    invalid_addresses = %w(user@example,com user_at_foo.org user.name@example. foo@bar_baz.com foo@bar+baz.com)
+    invalid_addresses.each do |invalid_address|
+      @user.email = invalid_address
+      @user.valid?
```
Comment (d-ray) [inline test/models/user_test.rb]: It might be a good to leave a comment here that calling `valid?` is necessary to populate `errors` on `@user`.  Otherwise the purpose of this line may not be obvious to future readers of the code.

**Comment by d-ray** [comment]
👍
